### ISSUE TITLE
Validation loss starts increasing after 20k steps, Tacotron2

### ISSUE THREAD

**Issue by bob80333** [espnet/espnet#1635]
I am using a custom ~10hr single male speaker dataset.  I used LJSpeech as the base, and trained using `train_pytorch_tacotron2.yaml`.  I expected it would take longer than 20k iterations to train, and the validation loss only gets to ~.48 before it starts to increase.  On Nvidia's tacotron2 repo I was able to get a validation loss of ~.2 when finetuning from their pretrained ljspeech model.

Here are my loss curves:
<img width="1106" alt="Screen Shot 2020-03-03 at 8 48 29 AM" src="https://user-images.githubusercontent.com/6364377/75781822-e8447980-5d2b-11ea-93ac-1987f58b7664.png">
<img width="1072" alt="Screen Shot 2020-03-03 at 8 48 22 AM" src="https://user-images.githubusercontent.com/6364377/75781823-e8447980-5d2b-11ea-82e1-4d77e658287a.png">

My alignments aren't amazing either:
![alignment5](https://user-images.githubusercontent.com/6364377/75781864-fb574980-5d2b-11ea-956c-944e78a653a7.png)
![alignment4](https://user-images.githubusercontent.com/6364377/75781865-fbefe000-5d2b-11ea-997c-cbf6d70c43f8.png)
![alignment3](https://user-images.githubusercontent.com/6364377/75781868-fc887680-5d2b-11ea-9a32-ea6513ce3e36.png)
![alignment2](https://user-images.githubusercontent.com/6364377/75781871-fc887680-5d2b-11ea-9acd-95c949a2b2bf.png)
![alignment1](https://user-images.githubusercontent.com/6364377/75781872-fc887680-5d2b-11ea-86a0-862f96b2c1d7.png)

Do you have any suggestions of hyperparameters to tweak or anything I could do to train a better model?

**Comment by CookiePPP** [comment]
To start things off
https://github.com/NVIDIA/tacotron2/blob/master/loss_function.py#L16
```        
        mel_loss = nn.MSELoss()(mel_out, mel_target) + \
            nn.MSELoss()(mel_out_postnet, mel_target)
```
NVIDIA/Tacotron2 uses MSE Loss, so I believe you should be comparing the mse_loss graph to Nvidia's implementation. (Please correct me if I'm misunderstanding this).
To further add fun for comparing the models. NVIDIA's masks like
https://github.com/NVIDIA/tacotron2/blob/master/model.py#L487
```
    def parse_output(self, outputs, output_lengths=None):
        if self.mask_padding and output_lengths is not None:
            mask = ~get_mask_from_lengths(output_lengths)
            mask = mask.expand(self.n_mel_channels, mask.size(0), mask.size(1))
            mask = mask.permute(1, 0, 2)

            outputs[0].data.masked_fill_(mask, 0.0)
            outputs[1].data.masked_fill_(mask, 0.0)
            outputs[2].data.masked_fill_(mask[:, 0, :], 1e3)  # gate energies
```
I don't believe masked_fill_ removes the values from the loss function, only makes all the padded for the output and target mels = 0.
ESPnet on the other hand uses 
https://github.com/espnet/espnet/blob/e860e7ca7bb8c780a49da8bee428cda2b1706ca5/espnet/nets/pytorch_backend/e2e_tts_tacotron2.py#L204

**Comment by kan-bayashi** [comment]
Thank you for your clarification. @CookiePPP! 
In terms of `mse` loss, your graph shows the model already achieved less than 0.2.

The difference between ours and nvidia is that 
- we use both `l1` and `mse` losses to make the attention learning faster
- we use guided attention loss to make the attention learning faster
Therefore, you should carefully compare the loss value, as @CookiePPP said.

If the length of the utterances in your dataset are varied, it is worthwhile to try `v4` config which uses `use-weighted-masking: true`. https://github.com/espnet/espnet/blob/master/egs/ljspeech/tts1/conf/tuning/train_pytorch_tacotron2.v4.yaml

**Comment by bob80333** [comment]
I did another training run with the v4 configuration, as you suggested.  

For some reason, I am noticing that the output of the model trained for 60-70k steps (~.52 total loss), which has a higher validation loss, sounds better / has more accurate pronunciation than the model with the lowest validation loss from ~20k training steps (~.48 total loss).

**Comment by kan-bayashi** [comment]
Right, in terms of perceptual quality, the validation loss best model is not the best. (This is one of the difficult issues of TTS...) Therefore, I recommend checking the quality even if the validation loss is increasing.
